### PR TITLE
Default dict arguments (with partial defaults)

### PR DESCRIPTION
**Pull request by idoby**
Add ability to pass a dictionary as the default value for a dataclass argument, with support for only some fields being provided a default value.

### CODE REVIEW

**Comment by idoby** [comment]
Hi @lebrice, I'm building an experiment automation library for ML and I'm using your library to pull args from the command line in a typesafe way. Great library, just missing a couple of things, which I want to add.

Thanks

File: test/test_default_args.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    class SomeClass:
+        a: some_type
+        b: int
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest="some_class", default={'a': expected_value})
+
+    args = parser.parse_args(shlex.split(f'--b 0'))
+    assert args == argparse.Namespace(some_class=SomeClass(a=expected_value, b=0))
+
```
Comment (lebrice) [inline test/test_default_args.py:78]: Could add a test for the other case, just to be safe:
```suggestion
    with raises_missing_required_arg():
        parser.parse_args(shlex.split("--b 123"))
```

File: test/test_default_args.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    @dataclass
+    class SomeClass:
+        a: some_type
+        b: int
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest="some_class", default={'a': expected_value})
+
+    args = parser.parse_args(shlex.split(f'--b 0'))
```
Comment (lebrice) [inline test/test_default_args.py:76]: nit:
```suggestion
    args = parser.parse_args(shlex.split("--b 0"))
```

File: test/test_default_args.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    some_type, passed_value, expected_value = simple_attribute
+
+    @dataclass
+    class SomeClass:
+        a: some_type
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest="some_class")
+
+    args = parser.parse_args(shlex.split(f'--a {passed_value}'))
```
Comment (lebrice) [inline test/test_default_args.py:16]: nit:
```suggestion
    args = parser.parse_args(shlex.split(f"--a {passed_value}"))
```

File: test/test_default_args.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    some_type, passed_value, expected_value = simple_attribute
+
+    @dataclass
+    class SomeClass:
+        a: some_type
+
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest="some_class", default=SomeClass(a=expected_value))
+
+    args = parser.parse_args('')
```
Comment (lebrice) [inline test/test_default_args.py:33]: nit:
```suggestion
    args = parser.parse_args("")
```

File: simple_parsing/wrappers/dataclass_wrapper.py [hunk 5]
```diff
@@ -12,19 +12,20 @@
...
+
 class DataclassWrapper(Wrapper[Dataclass]):
     def __init__(self,
                  dataclass: Type[Dataclass],
                  name: str,
-                 default: Dataclass=None,
-                 prefix: str="",
-                 parent: "DataclassWrapper"=None,
-                 _field: dataclasses.Field=None,
+                 default: Union[Dataclass, Dict] = None,
```
Comment (lebrice) [inline simple_parsing/wrappers/dataclass_wrapper.py:20]: Neat! Would you mind also changing the type annotation for the `default` argument of `ArgumentParser.add_arguments` ([found here](https://github.com/idoby/SimpleParsing/blob/34a33172c7235f72be1ec30f31e0741e376c0759/simple_parsing/parsing.py#L108)) as well? Not a big deal though.

**Review by lebrice — APPROVED** [review]
Hey @idoby , thanks a lot, this is a neat feature! There's ony very minor thing you could add, the rest of the changes can be added as a single batch from the GitHub UI.

Is your project open-source? If so, then I'd be curious to take a look at it!
I'd also be interested to hear more about the kind of features you'd like to add to SimpleParsing! 
Thanks again!

File: simple_parsing/wrappers/dataclass_wrapper.py (hunk 5 above)
Comment (idoby) [reply simple_parsing/wrappers/dataclass_wrapper.py:20]: Done. PyCharm did some extra formatting, which I left in this time. Feel free to revert the cosmetic changes

**Comment by idoby** [comment]
Hi, my project is not open source yet, it's too fresh for that yet.
The first thing I want to add is the ability to parse polymorphic data classes from the command line, but this will be tricky...

**Comment by lebrice** [comment]
Merging this now, thanks for your contribution!

Just out of curiosity, what do you mean by "parse polymorphic data classes"?

**Review by lebrice — APPROVED** [review]

**Comment by idoby** [comment]
Thanks for the merge, mate!

I mean getting something like this from the command line in a typesafe way:

```python
@dataclass 
class Base:
    num1: int = 5

@dataclass
class Derived1(Base):
    num_specific_to_this_class: float = 7.5

class Derived2(Base):
    ...
```

**Comment by lebrice** [comment]
This isn't hard at all though! It's already supported with Simple-Parsing. Depending on what you want to do, subparsers might be the best way to go about this. Let me know if you have any issues with this.

**Comment by idoby** [comment]
Hi @lebrice 
It seems like both features I added (this and #124) got marked as deprecated.
Do you intend to remove them? Why?
In that case, I will need to find an alternative to SP or fork it.

**Comment by lebrice** [comment]
Hey @idoby , If I'm being honest, I don't remember why exactly I marked those as deprecated. I'm not thinking of removing them anytime soon.

Feel free to fork the repo if you want, but I don't intend to remove those without good reason.

**Pull request merged**